python-package/xgboost/dask.py

```diff
@@ -136,11 +152,26 @@ async def __aexit__(self, *args: Any, **kwargs: Any) -> None:
     return MultiLock
 
 
-def _start_tracker(n_workers: int) -> Dict[str, Any]:
-    """Start Rabit tracker """
-    env: Dict[str, Union[int, str]] = {'DMLC_NUM_WORKER': n_workers}
-    host = get_host_ip('auto')
-    rabit_context = RabitTracker(hostIP=host, n_workers=n_workers, use_logger=False)
+def _start_tracker(
+    n_workers: int, addr_from_dask: Optional[str], addr_from_user: Optional[str]
+) -> Dict[str, Any]:
+    """Start Rabit tracker"""
+    env: Dict[str, Union[int, str]] = {"DMLC_NUM_WORKER": n_workers}
+    try:
+        rabit_context = RabitTracker(
+            hostIP=get_host_ip(addr_from_user), n_workers=n_workers, use_logger=False
+        )
+    except socket.error as e:
+        if e.errno != 99 or addr_from_dask is None:  # not a bind error
+            raise
+        LOGGER.warning(
+            "Failed to bind address '%s', trying to use '%s' instead.",
+            str(get_host_ip(addr_from_user)),
+            str(addr_from_dask),
+        )
+        rabit_context = RabitTracker(
```

Comment:
This statement is guarded in a try/catch block above.  Repeating it in the except clause exposes the module to a catastrophic exception.  Would it be better to make this function recursive on except with a base case that checks that addr_from_user does not equal addr_from_dask before recursing on addr_from_dask.?

Reply:
Thanks for the suggestion.  Let me try to factor out this part of the code.

Reply:
@kevingerman I converted it into a recursive function, but the exception is still raised and depends on dask to handle it.  I think throwing the socket error might be the best way to handle it, dask scheduler/workers will continue to run while xgboost training will be stopped.  What do you think?